mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/memory/RetentionTask.java

```diff
@@ -0,0 +1,133 @@
+        private final long permanantlyDeletedMessages;
+        private final long vaultSearchErrorCount;
+        private final long deletionErrorCount;
+
+        AdditionalInformation(Duration retentionPeriod, Optional<ZonedDateTime> beginningOfRetentionPeriod, long handledUserCount, long permanantlyDeletedMessages, long vaultSearchErrorCount, long deletionErrorCount) {
+            this.retentionPeriod = retentionPeriod;
+            this.beginningOfRetentionPeriod = beginningOfRetentionPeriod;
+            this.handledUserCount = handledUserCount;
+            this.permanantlyDeletedMessages = permanantlyDeletedMessages;
+            this.vaultSearchErrorCount = vaultSearchErrorCount;
+            this.deletionErrorCount = deletionErrorCount;
+        }
+
+        public Duration getRetentionPeriod() {
+            return retentionPeriod;
+        }
+
+        public  Optional<ZonedDateTime> getBeginningOfRetentionPeriod() {
+            return beginningOfRetentionPeriod;
+        }
+
+        public long getHandledUserCount() {
+            return handledUserCount;
+        }
+
+        public long getPermanantlyDeletedMessages() {
+            return permanantlyDeletedMessages;
+        }
+
+        public long getVaultSearchErrorCount() {
+            return vaultSearchErrorCount;
+        }
+
+        public long getDeletionErrorCount() {
+            return deletionErrorCount;
+        }
+    }
+
+    private static final String TYPE = "deletedMessageVaultRetention";
+
+    private final RetentionConfiguration configuration;
+    private final DeleteByQueryExecutor deleteByQueryExecutor;
+    private final Clock clock;
+    private final DeleteByQueryExecutor.Notifiers notifiers;
+    private final AtomicLong handledUserCount;
+    private final AtomicLong permanantlyDeletedMessages;
+    private final AtomicLong vaultSearchErrorCount;
+    private final AtomicLong deletionErrorCount;
+    private Optional<ZonedDateTime> beginningOfRetentionPeriod;
+
+    public RetentionTask(RetentionConfiguration configuration, DeleteByQueryExecutor deleteByQueryExecutor, Clock clock) {
+        this.configuration = configuration;
+        this.deleteByQueryExecutor = deleteByQueryExecutor;
+        this.clock = clock;
+
+        this.beginningOfRetentionPeriod = Optional.empty();
+
+        this.handledUserCount = new AtomicLong(0);
+        this.permanantlyDeletedMessages = new AtomicLong(0);
+        this.vaultSearchErrorCount = new AtomicLong(0);
+        this.deletionErrorCount = new AtomicLong(0);
+
+        this.notifiers = new DeleteByQueryExecutor.Notifiers(
+            handledUserCount::incrementAndGet,
+            vaultSearchErrorCount::incrementAndGet,
+            deletionErrorCount::incrementAndGet,
+            permanantlyDeletedMessages::incrementAndGet);
+    }
+
+    @Override
+    public Result run() {
+        ZonedDateTime now = ZonedDateTime.ofInstant(clock.instant(), ZoneOffset.UTC);
```

Comment:
why don't we compute it at build time?
If I ask for a GC the 25 of March at 12am and the retention is 1 day, I want message before 24 of March 12am to be deleted.
Now, imagine the task is run a 13am because the Task Manager is busy. We would remove message between 12am and 1pm. That also means that the task is not idempotent.

Reply:
By build time, do you mean `task submission time`?

Sounds like a pretty good idea...

Reply:
Héhé also solves the mutability problem! Thanks for the suggestion.